### PR TITLE
Enable users to create custom cpp op outside framework.

### PR DESCRIPTION
**Pull request by qingqing01**
Doc: https://github.com/PaddlePaddle/FluidDoc/pull/1443

1.  Package fluid_framework.so and headers into whl.
  - Package and release shared library  fluid_framework.so, which include core framework, exclude `paddle/fluid/operators` and `paddle/fluid/inference`.

  libs tree dir:   Total 142M

```
libs/  # 142M
|-- libfluid_framework.so  # 7.6M
|-- libiomp5.so
|-- libmklml_intel.so
`-- libwarpctc.so
```

  - Package headers include `paddle/fluid/framework`,  `paddle/fluid/memory`, `paddle/fluid/platform`、`paddle/fluid/string` and `third_party` into whl.

  include tree dir:  Total 75M

```
include/
|-- paddle # 2M
|   `-- fluid
|       |-- framework
|       |-- memory
|       |-- platform
|       `-- string
`-- third_party # 73M
    |-- boost  # 63M
    |   `-- boost
    |-- eigen3
    |   |-- Eigen
    |   `-- unsupported
    |-- install
    |   |-- gflags
    |   |-- glog
    |   |-- protobuf
    |   `-- xxhash
    `-- threadpool
        `-- src
```

Users can get include and lib dir by:

```python
paddle.sysconfig.get_include()
paddle.sysconfig.get_lib()
```

2.  Export some C-APIs to merge OpInfo between core.so and custom_op.so.

3.  How to write op needs to follow the op spec.

4. Add unit testing.

### CODE REVIEW

File: paddle/fluid/framework/c/c_api.cc [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+#include <vector>
+
+#include "paddle/fluid/framework/block_desc.h"
+#include "paddle/fluid/framework/c/c_api.h"
+#include "paddle/fluid/framework/op_registry.h"
+#include "paddle/fluid/framework/program_desc.h"
+#include "paddle/fluid/framework/scope.h"
+#include "paddle/fluid/platform/init.h"
+
+extern "C" {
```
Comment (wopeizl) [inline paddle/fluid/framework/c/c_api.cc:27]: 预测同学正在做C-API相关的事情，这里的API是否可以合到一起？

File: paddle/fluid/framework/load_op.h [hunk 2]
```diff
@@ -0,0 +1,115 @@
+/* Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (wopeizl) [inline paddle/fluid/framework/load_op.h]: 这个名字和operator下的load_op重名了，是不是容易混淆，可以考虑改一个名字

File: paddle/fluid/platform/dynload/dynamic_loader.cc [hunk 3]
```diff
@@ -264,6 +266,16 @@ void* GetMKLMLDsoHandle() {
 #endif
 }
 
+void* GetOpDsoHandle(const std::string& dso_name) {
+#if defined(__APPLE__) || defined(__OSX__)
+  PADDLE_THROW("Do not support Apple.");
+#elif defined(_WIN32) && defined(PADDLE_WITH_CUDA)
+  PADDLE_THROW("Do not support Windows.");
```
Comment (wopeizl) [inline paddle/fluid/platform/dynload/dynamic_loader.cc:289]: 不支持mac和win吗

File: python/paddle/fluid/tests/custom_op/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,28 @@
+message("fluid_framework_shared ${fluid_framework_shared}")
+if (WITH_GPU)
+    #nv_library(relu_op_shared SHARED relu_op.cc relu_op.cu DEPS fluid_framework_shared)
+    nv_library(relu_op_shared SHARED SRCS relu_op.cc relu_op.cu)
```
Comment (wopeizl) [inline python/paddle/fluid/tests/custom_op/CMakeLists.txt]: relu op加这里的原因是？

File: python/setup.py.in [hunk 5]
```diff
@@ -216,11 +222,21 @@ if '${WITH_NGRAPH}' == 'ON':
     package_data['paddle.libs']+=['${NGRAPH_SHARED_LIB_NAME}',
                                   '${NGRAPH_CPU_LIB_NAME}',
                                   '${NGRAPH_TBB_LIB_NAME}']
+
+# copy libfuild_framework.so to libs
+fluid_framework_lib='${FLUID_FRAMEWORK_SHARED_LIB}'
+print(fluid_framework_lib)
```
Comment (wopeizl) [inline python/setup.py.in]: 这行不需要

File: paddle/fluid/framework/CMakeLists.txt [hunk 6]
```diff
@@ -243,3 +243,35 @@ message(STATUS "commit: ${PADDLE_COMMIT}")
...
+cc_library(fluid_framework_shared
+    SHARED SRCS executor.cc operator.cc ${CMAKE_CURRENT_SOURCE_DIR}/c/c_api.cc
+    DEPS ${FLUID_FRAMEWORK_MODULES})
+get_property(os_dependency_modules GLOBAL PROPERTY OS_DEPENDENCY_MODULES)
+set_target_properties(fluid_framework_shared PROPERTIES OUTPUT_NAME fluid_framework)
+target_link_libraries(fluid_framework_shared ${os_dependency_modules})
+
+if (LINUX)
+  set(FLUID_FRAMEWORK_SHARED_LIB
+      ${PADDLE_BINARY_DIR}/paddle/fluid/framework/libfluid_framework.so
```
Comment (wopeizl) [inline paddle/fluid/framework/CMakeLists.txt]: 有一个libpaddle_fluid.so，是否可以直接用那个

File: python/paddle/fluid/tests/custom_op/CMakeLists.txt (hunk 4 above)
Comment (qingqing01) [reply python/paddle/fluid/tests/custom_op/CMakeLists.txt]: 这里是作为单测

File: paddle/fluid/platform/dynload/dynamic_loader.cc (hunk 3 above)
Comment (qingqing01) [reply paddle/fluid/platform/dynload/dynamic_loader.cc:289]: 由于还没在mac和win上验证，暂时关闭对mac和win的支持~

File: paddle/fluid/framework/CMakeLists.txt (hunk 6 above)
Comment (qingqing01) [reply paddle/fluid/framework/CMakeLists.txt]: 和libpaddle_fluid.so不同的是，libfluid_framework.so中不包含`paddle/fluid/operator`目录中的op，`paddle/fluid/operator`中注册了很多op，这样会导致load动态库的时候，这些已注册的op会再次load一次。

File: python/setup.py.in (hunk 5 above)
Comment (qingqing01) [reply python/setup.py.in]: Done.

File: paddle/fluid/framework/load_op.h (hunk 2 above)
Comment (qingqing01) [reply paddle/fluid/framework/load_op.h]: Done, 改成了 load_op_lib

**Review by sneaxiy — DISMISSED** [review]
LGTM.

**Review by heavengate — DISMISSED** [review]

**Comment by willthefrog** [comment]
Not sure it is a good idea to include all those third_party headers, which adds quite a bit  to the installed size. Maybe provide a demo cmake file that fetches third_party source automatically?
otherwise LGTM

**Review by sneaxiy — DISMISSED** [review]
LGTM.

**Review by sneaxiy — DISMISSED** [review]

**Comment by qingqing01** [comment]
```
[12:15:52]
[Step 1/1] 1 . You must have Dianhai approval for change 20+ files or add than 1000+ lines of content
[12:15:52]
[Step 1/1] 2 . You must have two RD (wanghaoshuang or guoshengCS or heavengate or kuke or Superjomn or lanxianghit or cyj1986 or hutuxian or frankwhzhang or nepeplwu) approval for the api change! paddle/fluid/API.spec for the management reason of API interface and API document.
[12:15:52]
[Step 1/1] 3 . You must have one RD (XiaoguangHu01,chengduoZH,Xreki,luotao1,sneaxiy,tensor-tang) approval for python/paddle/fluid/framework.py, which manages the underlying code for fluid.
```

**Review by sneaxiy — APPROVED** [review]

**Review by heavengate — APPROVED** [review]

**Review by raindrops2sea — APPROVED** [review]

**Review by guoshengCS — APPROVED** [review]

**Pull request merged**